tiledb/sm/storage_manager/storage_manager.cc

```diff
@@ -1242,11 +1144,49 @@ Status StorageManager::write(const URI& uri, Buffer* buffer) const {
   return vfs_->write(uri, buffer->data(), buffer->size());
 }
 
+Status StorageManager::xlock_array(const URI& array_uri) {
+  // Wait until the array is closed
+  std::unique_lock<std::mutex> lk(open_array_mtx_);
+  xlock_cv_.wait(lk, [this, array_uri] {
+    return open_arrays_.find(array_uri.to_string()) == open_arrays_.end();
+  });
+
+  // Retrieve filelock
+  filelock_t filelock = INVALID_FILELOCK;
+  auto lock_uri = array_uri.join_path(constants::filelock_name);
+  RETURN_NOT_OK(vfs_->filelock_lock(lock_uri, &filelock, false));
+  filelocks_[array_uri.to_string()] = filelock;
+  lk.unlock();
+
+  return Status::Ok();
+}
+
+Status StorageManager::xunlock_array(const URI& array_uri) {
+  open_array_mtx_.lock();
+
+  // Get filelock if it exists
+  auto it = filelocks_.find(array_uri.to_string());
+  if (it == filelocks_.end())
+    return LOG_STATUS(Status::StorageManagerError(
+        "Cannot unlock array exclusive lock; Filelock not found"));
+  auto filelock = it->second;
+
+  auto lock_uri = array_uri.join_path(constants::filelock_name);
+  if (filelock != INVALID_FILELOCK)
+    RETURN_NOT_OK(vfs_->filelock_unlock(lock_uri, filelock));
+  filelocks_.erase(it);
+
+  open_array_mtx_.unlock();
+  xlock_cv_.notify_all();
```

Comment:
Isn't it required to hold the associated lock before notifying a CV?

Reply:
Not based on [this example](https://en.cppreference.com/w/cpp/thread/condition_variable).